### DISCUSSION TITLE
General Talk in Japanese

### DISCUSSION THREAD

**Discussion by emadurandal** [actnwit/RhodoniteTS#930] [General]
# General Talk in Japanese

こちらは日本語OKです。Rhodonite全般について、質問や要望などを受け付けます。
今後、特定の話題（Q&Aや要望）が増えてきたら、専用のDiscussionを設置します。

**Comment by emadurandal** [comment]
Rhodonite v0.6.1がリリースされました。
v0.6.0と合わせて破壊的な変更が多く含まれます。

* "System", "GltfImporter", "Gltf2Importer", "Gltf1Importer", "VrmImporter", "ModelComverter"のシングルトンパターンが廃止され、staticなクラス（staticメンバ関数のみを持つクラス）になりました。
* "VRMImporter"クラスが"VrmImporter"にクラス名変更
* "webgl"や"pbr"などのLoadModuleが不要になりました。
* "system.setProcessApproachAndCanvas" メソッドは"System.init"メソッドになりました。使うときはawaitをつけるのを忘れないようにしてください。

ここ最近の変更で、だいぶコードが簡潔に書けるようになったと思います。

**Comment by emadurandal** [comment]
最新のmasterですがすでにPrimitive-basedになっています。例のビットフィールドベースの描画順ソートです。
https://github.com/actnwit/RhodoniteTS/issues/997
速くなってます。
https://twitter.com/emadurandal_cgj/status/1512719266080260102?s=21

あと、パッケージ処理関連を見直し、CommonJSとして利用する（npmインストールして使う方法）場合、型についてはこれまで以下のようにNamed Importする必要がありましたが、

```typescript
import Rn, {Entity, CubeTexture, HdriFormatEnum, Material, MeshRendererComponent, RenderPass} from 'rhodonite'
const entity: Entity = someEntity;
const textureCube: CubeTexture = someTexCube;
```

最新のmasterではRnという名前空間で型を使えます

```typescript
import Rn from 'rhodonite'
const entity: Rn.Entity = someEntity;
const textureCube: Rn.CubeTexture = someTexCube;
```

**Comment by emadurandal** [comment]
Rhodonite v0.7.0がリリースされました。
Matrix関連クラスやQuaternion関連クラスのコンストラクタがFloat32Arrayのみを受け付けるように単純化されました。

一般的にVectorやMatrix、Quaternionクラスのインスタンスを作成するには、"fromXXX"スタティックメソッドを使います。

```ts
const matrix1 = Matrix44.fromCopy16RowMajor(
  1, 0, 0, 5,
  0, 1, 0, 0,
  0, 0, 1, 0,
  0, 0, 0, 1,
);
const matrix2 = Matrix44.fromCopy16ColumnMajor(
  1, 0, 0, 0,
  0, 1, 0, 0,
  0, 0, 1, 0,
  5, 0, 0, 1,
);
const matrix3 = Matrix44.fromCopyQuaternion(quat);
const quat = Quaternion.fromCopy4(0, 0, 0, 1);
const vec = Vector4.fromCopy4(0, 0, 0, 1);
const vec2 = Vector4.fromCopyArray4([0, 0, 0, 1]);
```

**Reply by cx20** [reply]
現在、gltf-test の Rhodonite サンプルを `v0.6.3`→`v0.7.0` に変更中です。
v0.7.0 に変更した所、`Rn.EnvConstantSingleMaterialNode` が見当たりませんでした。下記に変更された、ということで良いでしょうか？

v0.6.3）Rn.EnvConstantSingleMaterialNode
v0.7.0）Rn.EnvConstantMaterialContent

**Reply by emadurandal** [reply]
@cx20 はい、名称変更されました。

**Comment by cx20** [comment]
現在、gltf-test の Rhodonite サンプルを `v0.7.1`→`v0.8.0` に変更中です。（ローカルにてテスト中）
v0.8.0 に変更した所、以前より暗くなってしまっているようです。使い方の問題でしょうか？

![image](https://user-images.githubusercontent.com/1833979/169718347-891fd0ba-b161-40aa-9492-b9a8826cdb70.png)

![image](https://user-images.githubusercontent.com/1833979/169718357-5d30febd-e7d8-4db9-9dfc-2f0aa7cbaf15.png)

**Reply by emadurandal** [reply]
@cx20 あ、ちょっとシェーダーを書き換えてですね。一部アプリケーションコード側の修正が必要です。コード見せていただけますか？

**Reply by emadurandal** [reply]
@cx20 今度Meetで一緒に見てみましょう。見当はついているので。

**Reply by cx20** [reply]
この辺りだとは思うのですが、どう直せば良いのか分かりませんでした。。また今度教えて下さい。
https://github.com/cx20/gltf-test/blob/c43207f9b03237d30c2ee5594ed9d38e1453d3b8/examples/rhodonite/index.js#L120-L134

**Reply by emadurandal** [reply]
@cx20 https://github.com/cx20/gltf-test/blob/c43207f9b03237d30c2ee5594ed9d38e1453d3b8/examples/rhodonite/index.js#L238

の

```ts
    sphereMaterial.setParameter(Rn.EnvConstantMaterialContent.EnvHdriFormat, Rn.HdriFormat.LDR_LINEAR.index);
```

を

```ts
    sphereMaterial.setParameter(Rn.ShaderSemantics.EnvHdriFormat, Rn.HdriFormat.LDR_LINEAR.index);
```

に変えてみてもらえますでしょうか。仕様変更なんでこれはわからなくてもしょうがないです😅

**Reply by cx20** [reply]
ありがとうございます。解消できました！
https://cx20.github.io/gltf-test/examples/rhodonite/index.html?category=sampleModels&model=Duck&scale=1&type=glTF
![image](https://user-images.githubusercontent.com/1833979/170307746-b584a848-c121-4d16-956c-4348e1300a87.png)

これは聞かないと分からなかったかもしれませんね。

**Comment by emadurandal** [comment]
Rhodonite v0.8.0をリリースしました。これまでのCommonJS版に加え、ESModule版を追加しています。サンプルコードもESModule版を使うように書き換えました。

**Comment by cx20** [comment]
現在、gltf-test の Rhodonite サンプルを `v0.8.0`→`v0.8.3` に変更中です。
透明になってくれないのですが、これも使い方の問題でしょうか？（どこを直してよいのか分からず・・）

![image](https://user-images.githubusercontent.com/1833979/171498262-8a3d3702-41f2-4f26-9e07-1db60cf14058.png)

![image](https://user-images.githubusercontent.com/1833979/171498334-cdb05113-637b-4a44-876a-e663dd86eebc.png)

**Reply by cx20** [reply]
Rhodonite Editor は良さそうですね。
![image](https://user-images.githubusercontent.com/1833979/179345961-8b566c64-c57b-4181-ac74-2ddc5b12b152.png)

**Reply by cx20** [reply]
gltf-test のコードだと ToyCar のフロントガラスが透過しないのですが、同様の事象でしょうか？
![image](https://user-images.githubusercontent.com/1833979/179346049-37702f90-165e-4ab0-8c1a-62fb899434d1.png)

Rhodonite Editor は良さそうです。
![image](https://user-images.githubusercontent.com/1833979/179346059-d97749c4-07f0-42ae-8d2c-26b4d6666a16.png)

**Comment by emadurandal** [comment]
Rhodonite v0.9.2をリリースしました。

KHR_materials_iridescenceやKHR_metarials_variantsに対応しました。
glTFImpoterなどの戻り値の型が変わっているので、利用側のコードの改修が必要です。

**Comment by cx20** [comment]
そういえば、glTF の仕様としては複数のシーンを含めることができるようです。
仕様としては存在していましたが、それを実装したサンプルが存在していなかったようで、最近サンプルが追加されたようです。
下記サンプルには四角形と三角形がそれぞれのシーンで提供されています。想定では四角形が表示されるはずですがRhodoniteだと三角形が表示されるようです。

https://github.com/javagl/glTF-Sample-Models/tree/MultipleScenes/2.0/MultipleScenes

https://cx20.github.io/gltf-test/examples/rhodonite/index.html?category=tutorialModels&model=MultipleScenes&scale=1&type=glTF

**Reply by emadurandal** [reply]
@cx20 お、それは知りませんでした。確認しますね。

**Comment by emadurandal** [comment]
Rhodonite v0.10.2をリリースしました。

ESModule版の拡張子を.mjsから.jsに変更したことで、ESModule版を使用したときにVS Codeで型チェックがきちんと働くようになりました（samplesフォルダ以下のサンプルがESModule版を利用しています）

**Reply by emadurandal** [reply]
それと、Rhodoniteのデフォルトブランチがmasterからmainに変更になりました。

**Comment by emadurandal** [comment]
Rhodonite v0.11.0をリリースしました。

以下の破壊的変更があります。
TransformComponentの以下のプロパティの名前が変更になりました。
* translate -> localPosition
* rotate -> localEulerAngles
* scale -> localScale
* quaternion -> localRotation
* matrix -> localMatrix

SceneGraphComponentの以下のプロパティの名前が変更になりました。
* translate -> position
* rotate -> eulerAngles
* quaternion -> rotation
* worldMatrix -> matrix

上記はUnityとPlayCanvasにおけるTransformの命名規則であり、それらに合わせた変更となります。

**Comment by cx20** [comment]
Wiki の イントロダクション にあるコードがエラーになるようです。
https://github.com/actnwit/RhodoniteTS/wiki/%E3%82%A4%E3%83%B3%E3%83%88%E3%83%AD%E3%83%80%E3%82%AF%E3%82%B7%E3%83%A7%E3%83%B3#rhodonite%E3%81%93%E3%81%A8%E3%81%AF%E3%81%98%E3%82%81

正しくは下記でしょうか？

誤）system.processAuto();
正）Rn.System.processAuto();

**Reply by emadurandal** [reply]
@cx20 ありがとうございます。直しました。

**Comment by emadurandal** [comment]
Rhodonite v0.11.2をリリースしました。

* EntityRepositoryクラスにEntityのコピーを作成するshallowCopyEntityメソッドを追加しました。
* shallowCopyEntityメソッドを使ってInstanced Skin Meshが機能するようになりました。

https://github.com/actnwit/RhodoniteTS/releases/tag/v0.11.2

**Reply by emadurandal** [reply]
@cx20 あと、MeshHelperの各種create関数の引数を部分省略できるようになりました。
